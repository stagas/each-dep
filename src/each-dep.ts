import * as fs from 'fs/promises'
import * as path from 'path'

const matchers =
  /((im|ex)port.*?from\s+['"](?<id1>[^'"]+)['"])|(import\(?\s*['"](?<id2>[^'"]+)['"])|(require\(\s*['"](?<id3>[^'"]+)['"])/g

const parseIds = (x: string) =>
  [...x.matchAll(matchers)]
    .map(x => x.groups?.id1 || x.groups?.id2 || x.groups?.id3).filter(Boolean) as string[]

const extensions = [
  '',
  '.ts',
  '.mjs',
  '.js',
  '.tsx',
  '.jsx',
  '.json',
  '.mts',
  '.mtsx',
  '.mjsx',
]

export interface EachDepOptions {
  cache?: Map<string, { source: string; ids: (readonly [string, string])[] }>
  external?: boolean
}

export async function* eachDep(
  inputFilename: string,
  options: EachDepOptions = {},
  seen = new Set<string>(),
): AsyncGenerator<string, void, undefined> {
  const resolved = path.resolve(inputFilename)
  const parsed = path.parse(resolved)

  yield resolved
  seen.add(resolved)

  try {
    let ids: (readonly [string, string])[] = []
    let source: string

    if (options.cache && options.cache.has(resolved)) {
      ;({ source, ids } = options.cache.get(resolved)!)
    } else {
      source = await fs.readFile(resolved, 'utf-8')

      if (options.cache) {
        options.cache.set(resolved, { source, ids })
      }

      ids.push(...(await Promise.all(
        parseIds(source).map(async x => {
          if (x.startsWith('.')) {
            const joined = path.join(parsed.dir, x)
            for (const ext of extensions) {
              const filename = `${joined}${ext}`
              try {
                const stat = await fs.stat(filename)
                if (stat.isFile()) return [x, filename] as const
                else if (stat.isDirectory()) {
                  for (const indexExt of extensions) {
                    const index = path.join(filename, `index${indexExt}`)
                    try {
                      const stat = await fs.stat(index)
                      if (stat.isFile()) return [x, index] as const
                    } catch {}
                  }
                }
              } catch {}
            }
          } else {
            if (options.external) {
              try {
                const importMetaResolve = (await eval('import(\'import-meta-resolve\')')).resolve
                const result = await importMetaResolve(x, `file://${resolved}`)
                const { pathname } = new URL(result)
                if (pathname.startsWith('/')) return [x, pathname] as const
              } catch {}
            }
          }
        })
      )).filter(Boolean) as (readonly [string, string])[])
    }

    const unseen = ids.filter(([, x]) => {
      if (!seen.has(x)) {
        seen.add(x)
        return x
      }
    })

    if (unseen.length) {
      for (const [, id] of unseen) yield* eachDep(id, options, seen)
    }
  } catch (error) {
    console.error(error)
  }
}
