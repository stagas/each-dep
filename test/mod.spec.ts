// @env node
import * as path from 'path'
import { eachDep } from '../src/each-dep'

describe('eachDep', () => {
  it('works', async () => {
    let deps = []
    for await (const dep of eachDep(__filename)) {
      deps.push(dep)
    }
    deps = deps.map(x => path.relative(__dirname, x))
    expect(deps).toMatchSnapshot()
  })

  it('complex', async () => {
    let deps = []
    for await (const dep of eachDep(path.join(__dirname, 'fixture', 'a.ts'))) {
      deps.push(dep)
    }
    deps = deps.map(x => path.relative(__dirname, x))
    expect(deps).toMatchSnapshot()
  })

  it('dot', async () => {
    let deps = []
    for await (const dep of eachDep(path.join(__dirname, 'fixture', 'dot', 'dot.js'))) {
      deps.push(dep)
    }
    deps = deps.map(x => path.relative(__dirname, x))
    expect(deps).toMatchSnapshot()
  })

  it('with external', async () => {
    let deps = []
    for await (const dep of eachDep(path.join(__dirname, 'fixture', 'pkg-a', 'index.js'), { external: true })) {
      deps.push(dep)
    }
    deps = deps.map(x => path.relative(__dirname, x))
    expect(deps).toMatchSnapshot()
  })

  it('cache', async () => {
    let deps = []
    const cache = new Map()
    for await (const dep of eachDep(path.join(__dirname, 'fixture', 'pkg-a', 'index.js'), { external: true, cache })) {
      deps.push(dep)
    }
    deps = deps.map(x => path.relative(__dirname, x))
    expect(deps).toMatchSnapshot()
    expect([...cache]).toMatchSnapshot()
  })
})
