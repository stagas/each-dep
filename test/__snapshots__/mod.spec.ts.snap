
exports[`eachDep works 1`] = `
Array [
  "mod.spec.ts",
  "../src/each-dep.ts",
]
`;

exports[`eachDep complex 1`] = `
Array [
  "fixture/a.ts",
  "fixture/b.ts",
  "fixture/c/index.ts",
  "fixture/c/d.js",
]
`;

exports[`eachDep dot 1`] = `
Array [
  "fixture/dot/dot.js",
  "fixture/dot/index.js",
]
`;

exports[`eachDep with external 1`] = `
Array [
  "fixture/pkg-a/index.js",
  "fixture/pkg-b/index.js",
]
`;

exports[`eachDep cache 1`] = `
Array [
  "fixture/pkg-a/index.js",
  "fixture/pkg-b/index.js",
]
`;

exports[`eachDep cache 2`] = `
Array [
  Array [
    "/home/stagas/work/stagas/each-dep/test/fixture/pkg-a/index.js",
    Object {
      "ids": Array [
        Array [
          "pkg-b",
          "/home/stagas/work/stagas/each-dep/test/fixture/pkg-b/index.js",
        ],
      ],
      "source": "export * from 'pkg-b'
",
    },
  ],
  Array [
    "/home/stagas/work/stagas/each-dep/test/fixture/pkg-b/index.js",
    Object {
      "ids": Array [
        Array [
          "pkg-a",
          "/home/stagas/work/stagas/each-dep/test/fixture/pkg-a/index.js",
        ],
      ],
      "source": "export * from 'pkg-a'
",
    },
  ],
]
`;
